File: wick/step_3_filter.py
"""
Step 3: Contraction Filter

For the Fermi vacuum with vacuum bra <0| and ket |0>, only fully-contracted
strings (no remaining operators) survive.  All other terms vanish.
"""

from __future__ import annotations
from typing import List
from .term import Term


def filter_fully_contracted(terms: List[Term]) -> List[Term]:
    """Keep only terms with no remaining uncontracted operators."""
    return [t for t in terms if t.is_fully_contracted()]

File: src/step_3_filter.rs
//! Step 3: Contraction Filter
//!
//! Keep only fully-contracted terms (no remaining operators).

use crate::term::Term;

pub fn filter_fully_contracted(terms: &[Term]) -> Vec<Term> {
    terms
        .iter()
        .filter(|t| t.is_fully_contracted())
        .cloned()
        .collect()
}
